lib/pbench/server/database/models/users.py

```diff
@@ -0,0 +1,116 @@
+import datetime
+from flask_bcrypt import generate_password_hash
+from sqlalchemy import Column, Integer, String, DateTime, LargeBinary
+from pbench.server.database.database import Database
+from sqlalchemy.orm import relationship
+
+
+class User(Database.Base):
+    """ User Model for storing user related details """
+
+    __tablename__ = "users"
+
+    id = Column(Integer, primary_key=True, autoincrement=True)
+    username = Column(String(255), unique=True, nullable=False)
+    first_name = Column(String(255), unique=False, nullable=False)
+    last_name = Column(String(255), unique=False, nullable=False)
+    password = Column(LargeBinary(128), nullable=False)
+    registered_on = Column(DateTime, nullable=False)
+    email = Column(String(255), unique=True, nullable=False)
+    auth_tokens = relationship("ActiveTokens", backref="users")
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.username = kwargs.get("username")
```

Comment:
What ensures that a `User()` is not invoked with an empty `kwargs`?  Shouldn't we be requiring some arguments?

Reply:
@portante The only place where we invoke the `User()` class is during the user registration. Anywhere else you see the `user` object is a result of sqlalchemy query. Therefore I think the `kwargs` is fine in this case because it gets invoked only at 1 place and  its in our control which arguments we provide to the user class when user is registering himself.

Reply:
Not sure I can articulate exactly what makes me uneasy about this, but it is along the lines that the language constructs are not being used to help us ensure we get it right.

In particular, all of the `.get()` calls off of `kwargs` will silently return `None` if the parameter does not exist.

Further, the fields defined from lines 16 - 21 seem to be over-ridden here.  I would have thought the we can't do that assignment.  But I am probably missing the magic of it all.

I don't think this comment should hold up this commit, but would to see this turned into an two issues to ge investigated: the use of `kwargs` here, and what appears to be over-riding the class fields.

Reply:
Yes, that's how SQLAlchemy works: there are static Class attributes defining the columns of the database table, and Object attributes of the same name describing the values of those columns for a row. There are ways to define a database functionally rather than with the declarative class model, but I find that much less appealing.

Clients that really need to understand the DB schema defined by the model can get the information elsewhere in the API, and normal clients of the object interface only deal with the column values in the rows they're manipulating through the OO model.

Yeah, I suppose the occluded namespace could be considered a bit odd, but I don't think it's a problem, and this model is widely used.

Reply:
I entirely removed the `__init__` block so we dont need `kwargs`